File: elixir/aoc/lib/aoc/day03b.ex
defmodule AOC.Day03b do
  use AOC

  def solution(path) do
    path
    |> read_lines()
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.split(&1, "", trim: true))
    |> create_grid_map()
    |> find_gear_ratios()
    |> Enum.sum()
  end

  def find_gear_ratios(grid_map) do
    part_numbers = find_part_numbers(grid_map)

    grid_map
    |> Enum.filter(&asterisk_cell?/1)
    |> Enum.map(&parts_for_asterisk(&1, part_numbers))
    |> Enum.filter(fn parts -> length(parts) == 2 end)
    |> Enum.map(fn [part1, part2] -> part1 * part2 end)
  end

  def asterisk_cell?({_, "*"}), do: true
  def asterisk_cell?(_), do: false

  def parts_for_asterisk({coordinate, _}, part_numbers) do
    part_numbers
    |> Enum.filter(&part_close_to_coordinate?(&1, coordinate))
    |> Enum.map(fn {number, _} -> String.to_integer(number) end)
  end

  def part_close_to_coordinate?({_number, coordinates}, coordinate) do
    coordinates
    |> Enum.any?(&close_to_coordinate?(&1, coordinate))
  end

  def close_to_coordinate?({x1, y1}, {x2, y2}) do
    abs(x1 - x2) <= 1 and abs(y1 - y2) <= 1
  end

  def create_grid_map(grid) do
    grid
    |> Enum.with_index()
    |> Enum.map(&map_coordinates_to_characters/1)
    |> List.flatten()
    |> Map.new()
  end

  def map_coordinates_to_characters({row, y}) do
    row
    |> Enum.with_index()
    |> Enum.map(fn {character, x} -> {{x, y}, character} end)
  end

  def find_part_numbers(grid_map) do
    grid_map
    |> find_numbers()
    |> Enum.filter(&part_number?(&1, grid_map))
  end

  def part_number?({_number, coordinates}, grid_map) do
    coordinates
    |> Enum.any?(&near_symbol(&1, grid_map))
  end

  def near_symbol({x, y}, grid_map) do
    [
      {x - 1, y},
      {x + 1, y},
      {x, y - 1},
      {x, y + 1},
      {x - 1, y - 1},
      {x + 1, y + 1},
      {x - 1, y + 1},
      {x + 1, y - 1}
    ]
    |> Enum.any?(&symbol?(&1, grid_map))
  end

  def symbol?(coordinate, grid_map) do
    character = Map.get(grid_map, coordinate)

    cond do
      character == nil -> false
      character == "." -> false
      String.match?(character, ~r/\d/) -> false
      true -> true
    end
  end

  def find_numbers(grid_map) do
    grid_map
    |> Enum.filter(&number_cell?/1)
    |> Enum.sort(&sort_by_coordinate/2)
    |> Enum.chunk_while({}, &number_chunker/2, &number_finalizer/1)
  end

  def number_cell?({_, character}) do
    String.match?(character, ~r/\d/)
  end

  def sort_by_coordinate({{x, y}, _}, {{x, y}, _}), do: true
  def sort_by_coordinate({{_, y1}, _}, {{_, y2}, _}) when y1 < y2, do: true
  def sort_by_coordinate({{x1, y}, _}, {{x2, y}, _}) when x1 <= x2, do: true
  def sort_by_coordinate(_, _), do: false

  def number_chunker({coordinate, character}, {}), do: {:cont, {character, [coordinate]}}

  def number_chunker({{x1, y} = coordinate, character}, {number, [{x2, y} | _] = coordinates})
      when x1 - x2 == 1 do
    {:cont, {number <> character, [coordinate | coordinates]}}
  end

  def number_chunker({coordinate, character}, {number, coordinates}) do
    {:cont, {number, coordinates}, {character, [coordinate]}}
  end

  def number_finalizer({}), do: {:cont, {}}
  def number_finalizer({number, coordinates}), do: {:cont, {number, coordinates}, {}}
end

File: ruby/aoc/lib/aoc/day03b.rb
module AOC
  class Day03b < Day
    def sum_gear_ratios
      gear_ratios.sum
    end

    def gear_ratios
      @gear_ratios ||= grid_map.select(&method(:asterisk?))
        .map(&method(:parts_for_asterisk))
        .select {|parts| parts.length == 2}
        .map(&method(:gear_ratio))
    end

    def asterisk?(_coordinate, character)
      character == "*"
    end

    def parts_for_asterisk(asterisk)
      coordinate, _ = asterisk
      part_numbers.select {|part| part_number_near_coordinate?(part, coordinate)}
        .map do |part|
          part_number, _coordinates = part
          part_number.to_i
        end
    end

    def part_number_near_coordinate?(part, coordinate)
      _part_number, coordinates = part
      coordinates.any? {|part_coordinate| near_coordinate?(part_coordinate, coordinate)}
    end

    def near_coordinate?(part_coordinate, coordinate)
      x1, y1 = part_coordinate
      x2, y2 = coordinate
      (x1 - x2).abs <= 1 && (y1 - y2).abs <= 1
    end

    def gear_ratio(parts)
      part1, part2 = parts
      part1 * part2
    end

    def part_numbers
      @part_numbers ||= numbers.select(&method(:part_number?))
    end

    def part_number?(number)
      _, coordinates = number
      coordinates.any?(&method(:near_symbol?))
    end

    def near_symbol?(coordinate)
      x, y = coordinate
      [[x-1,y],[x+1,y],[x,y-1],[x,y+1],[x+1,y+1],[x-1,y+1],[x+1,y-1],[x-1,y-1]].any?(method(:symbol?))
    end

    def symbol?(coordinate)
      return false if grid_map[coordinate].nil?
      return false if grid_map[coordinate] == "."
      return false if grid_map[coordinate] =~ /\d/
      true
    end

    def numbers
      @numbers ||= grid_map.select(&method(:number?))
        .to_a
        .sort(&method(:cell_sorter))
        .chunk_while(&method(:contiguous_digits?))
        .map(&method(:number_with_coordinates))
    end

    def number_with_coordinates(chunk)
      chunk.reduce(["", []]) do |memo, cell|
        coordinate, digit = cell
        number, coordinates = memo
        [number + digit, coordinates.push(coordinate)]
      end
    end

    def contiguous_digits?(cell1, cell2)
      coordinate1, _ = cell1
      coordinate2, _ = cell2
      x1, y1 = coordinate1
      x2, y2 = coordinate2
      y1 == y2 && x2 - x1 == 1
    end

    def cell_sorter(cell1, cell2)
      coordinate1, _ = cell1
      coordinate2, _ = cell2
      x1, y1 = coordinate1
      x2, y2 = coordinate2
      return 0 if x1 == x2 && y1 == y2
      return -1 if y1 < y2
      return -1 if y1 == y2 && x1 < x2
      1
    end

    def number?(_coordinate, character)
      character =~ /\d/
    end

    def grid_map
      return @grid_map if @grid_map
      map = {}
      grid.each_with_index do |row, y|
        row.each_with_index { |cell, x| map[[x, y]] = cell }
      end
      @grid_map = map
    end

    def grid
      @grid ||= File.readlines(@path, chomp: true)
        .map { |line| line.split("") }
    end
  end
end
